README.md
# sliver_grouped_list

SliverGroupedList Widget

Build a beautiful sliver with a custom header, footer and body, which can be divided by groups with a header for each group and a list of entries.

<img src="assets/demo.gif" width="250" height="400"/>

## Getting Started

- Go inside pubspec.yaml file
- Add this to your package's pubspec.yaml file: 
``` Dart
dependencies:
     sliver_grouped_list: ^1.0.4
```
- Run <b>flutter pub get</b>

## How to use

Go to your widget .dart file and do import
``` Dart
import 'package:sliver_grouped_list/sliver_grouped_list.dart';
```
Inside your widget's <b>Widget build(BuildContext context)</b> method
set as body the <b>SliverGroupedList</b> with explicitly showing the types of bodyHeader and bodyEntry.
``` Dart
// [Header] is the type of a group header
// [Entry] is the type of every entry in the list related to the group
body: SliverGroupedList<Header, Entry>()
```
Implement all required constructor parameters.

As a result the custom scroll view will contain the content of the passed data.
Optionally you may add a <b>RenderObjectWidget</b> widget like <b>SliverGrid</b> or <b>SliverList</b> header section to the scroll view, and a footer as well.

Here is an example which is displayed in the GIF from above.

``` Dart
import 'package:flutter/material.dart';
import 'package:sliver_grouped_list/sliver_grouped_list.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  static const double _kHeight = 150.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example'),
      ),
      body: SliverGroupedList<String, String>(
        data: {
          'A': ['My best friend', 'Good friend of mine', 'Guy I do not know'],
          'B': ['My cat', 'My dog', 'My fish', 'My bird'],
          'C': ['My mom', 'My dad', 'My sis']
        },
        header: SliverGrid.count(
          crossAxisCount: 3,
          children: [
            Container(color: Colors.white, height: _kHeight),
            Container(color: Colors.black, height: _kHeight),
            Container(color: Colors.grey, height: _kHeight),
            Container(color: Colors.grey, height: _kHeight),
            Container(color: Colors.white, height: _kHeight),
            Container(color: Colors.black, height: _kHeight),
            Container(color: Colors.black, height: _kHeight),
            Container(color: Colors.grey, height: _kHeight),
            Container(color: Colors.white, height: _kHeight),
          ],
        ),
        bodyHeaderMinHeight: 60.0,
        bodyHeaderMaxHeight: 100.0,
        bodyHeaderPinned: true,
        bodyHeaderFloating: false,
        bodyHeaderBuilder: (_, header) => Container(
          alignment: Alignment.center,
          child: Text(
            header,
            style: TextStyle(color: Colors.white),
          ),
          color: Colors.blue,
        ),
        bodyPlaceholderBuilder: (_, header) => Card(
              color: Colors.lightBlueAccent,
              child: Container(
                  height: _kHeight,
                  alignment: Alignment.center,
                  child: Text(
                    "There are no items available",
                    style: TextStyle(color: Colors.grey),
                  ))),
        bodyEntryBuilder: (_, index, item) => GestureDetector(
          onTap: () {
            print(item);
            print(index);
          },
          child: Card(
              color: Colors.lightBlueAccent,
              child: Container(
                  height: _kHeight,
                  alignment: Alignment.center,
                  child: Text(
                    item,
                    style: TextStyle(color: Colors.white),
                  ))),
        ),
        footer: SliverFixedExtentList(
          itemExtent: _kHeight,
          delegate: SliverChildListDelegate(
            [
              Container(color: Colors.black),
              Container(color: Colors.white),
              Container(color: Colors.grey),
              Container(color: Colors.black),
              Container(color: Colors.white),
              Container(color: Colors.grey),
            ],
          ),
        ),
      ),
    );
  }
}
```
## License
Under <a href=https://github.com/VictorKachalov/sliver_grouped_list/blob/master/LICENSE>MIT License</a>

## Donation
If you think that my repo helped you to solve the issues you struggle with, please don't be shy and donate :-)

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/paypalme/VictorKachalov/5EUR)


CHANGELOG.md
## [1.1.2-nullsafety.0] - 10.04.2021.

* Add _bodyPlaceholderBuilder_
* Null safety support

## [1.1.1] - 05.10.2020.

* SliverAppBar support

## [1.1.0] - 04.10.2020.

* Removed GestureDetector. Should be implemented on customer side.
* Provide index in _bodyEntryBuilder_ 

## [1.0.5] - 28.08.2020.

* Better documentation.

## [1.0.4] - 28.08.2020.

* More properties to set up for a group's body entry header.

## [1.0.3] - 20.08.2020.

* Provide a valid description in pubspec.yaml file.

## [1.0.2] - 20.08.2020.

* Provide example.

## [1.0.1] - 20.08.2020.

* Make SliverList instead SliverFixedExtentList.

## [1.0.0] - 19.08.2020.

* Initial release. The Widget for making a grouped SliverList with Header/Entry content.

lib/sliver_grouped_list.dart
import 'dart:math';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

const double _kMaxHeightDim = 50.0;
const double _kMinHeightDim = 40.0;

/// A [SliverGroupedList] list which wraps a [CustomScrollView]
/// providing created [_SliverGroupedHeader] and [_SliverGroupedEntry] slivers
///
/// [Builder] is used to create a list of [Widget] and set
/// is as a slivers property of [CustomScrollView]
///
/// For every [_SliverGroupedHeader] and [_SliverGroupedEntry] a separate
/// builder method is created: [bodyHeaderBuilder] and [bodyEntryBuilder]
/// if no entries, [bodyPlaceholderBuilder] is created
///
/// [data] is based on provided [Header] and [Entry] types.
///
/// To build a [SliverGroupedList] the [Header] and [Entry] types
/// should be explicitly showed like [SliverGroupedList<String, String>]
///
/// Every item of the sliver list handles [onItemTap] event providing
/// the tapped item as a parameter.
class SliverGroupedList<Header, Entry> extends StatelessWidget {
  const SliverGroupedList(
      {Key? key,
      required this.data,
      this.bodyHeaderPinned = false,
      this.bodyHeaderFloating = false,
      this.bodyHeaderMinHeight = _kMinHeightDim,
      this.bodyHeaderMaxHeight = _kMaxHeightDim,
      required this.bodyHeaderBuilder,
      required this.bodyEntryBuilder,
      required this.bodyPlaceholderBuilder,
      this.appBar,
      this.header,
      this.footer,
      this.controller,
      this.primary,
      this.physics,
      this.center,
      this.cacheExtent,
      this.semanticChildCount,
      this.scrollDirection = Axis.vertical,
      this.reverse = false,
      this.shrinkWrap = false,
      this.anchor = 0.0})
      : super(key: key);

  /// Set [bodyHeaderMinHeight] and [bodyHeaderMaxHeight]
  /// to make collapsable header for entry
  /// [bodyHeaderMinHeight] is 40.0 by default
  /// [bodyHeaderMaxHeight] is 50.0 by default
  final double bodyHeaderMinHeight;
  final double bodyHeaderMaxHeight;

  /// Set [bodyHeaderPinned] to pin header and
  /// to enable/prevent dismissing by scrolling
  /// by default is set to false
  final bool bodyHeaderPinned;

  /// Set [bodyHeaderFloating] to
  /// toggle visibility of top header of the first group
  /// of elements
  /// by default is set to false
  final bool bodyHeaderFloating;

  /// [bodyHeaderBuilder] is mandatory and should not be null
  /// assertion is used to check the nullability of the
  /// property and will show a stacktrace
  final Widget Function(BuildContext context, Header header) bodyHeaderBuilder;

  /// [bodyEntryBuilder] is mandatory and should not be null
  /// assertion is used to check the nullability of the
  /// property and will show a stacktrace
  final Widget Function(BuildContext context, int index, Entry item)
      bodyEntryBuilder;

  /// [bodyPlaceholderBuilder] will be used when no elements are available
  /// for a certain key
  final Widget Function(BuildContext context, Header header)
      bodyPlaceholderBuilder;

  /// [data] is mandatory and should not be null
  /// assertion is used to check the nullability of the
  /// property and will show a stacktrace
  final Map<Header, List<Entry>> data;

  /// [appBar] provide additional Sliver related widget to
  /// display content above header as an app bar.
  final Widget? appBar;

  /// [header] provide additional Sliver related widget to
  /// display content above body.
  final RenderObjectWidget? header;

  /// [footer] provide additional Sliver related widget to
  /// display content under body.
  final RenderObjectWidget? footer;

  /// Properties of [CustomScrollView]
  final Axis scrollDirection;
  final bool reverse;
  final ScrollController? controller;
  final bool? primary;
  final ScrollPhysics? physics;
  final bool shrinkWrap;
  final Key? center;
  final double anchor;
  final double? cacheExtent;
  final int? semanticChildCount;
  final DragStartBehavior dragStartBehavior = DragStartBehavior.start;

  @override
  Widget build(BuildContext context) {
    return Builder(builder: (context) {
      var widgetList = <Widget>[];
      if (appBar != null) {
        widgetList.add(appBar!);
      }
      if (header != null) {
        widgetList.add(header!);
      }
      data.forEach((key, value) {
        widgetList
          ..add(_SliverGroupedHeader(
              minHeight: bodyHeaderMinHeight,
              maxHeight: bodyHeaderMaxHeight,
              pinned: bodyHeaderPinned,
              floating: bodyHeaderFloating,
              child: bodyHeaderBuilder(context, key)))
          ..add(value.length > 0
              ? _SliverGroupedEntry(entry: value, builder: bodyEntryBuilder)
              : SliverToBoxAdapter(
                  child: Builder(
                      builder: (context) =>
                          bodyPlaceholderBuilder(context, key))));
      });
      if (footer != null) {
        widgetList.add(footer!);
      }
      return CustomScrollView(
        scrollDirection: scrollDirection,
        reverse: reverse,
        controller: controller,
        primary: primary,
        physics: physics,
        shrinkWrap: shrinkWrap,
        center: center,
        anchor: anchor,
        cacheExtent: cacheExtent,
        slivers: widgetList,
        semanticChildCount: semanticChildCount,
        dragStartBehavior: dragStartBehavior,
      );
    });
  }
}

/// A [SliverFixedExtentList] list wrapper
///
/// Awaits for generic entry [Entry] for preparing
/// an item [builder]
/// Inside [builder] you can create your won [Widget] which
/// will be recognized as a list item with the ability
/// to call [onItemTap] callback if [GestureDetector] detects
/// onTap event
class _SliverGroupedEntry<Entry> extends StatelessWidget {
  const _SliverGroupedEntry({
    Key? key,
    required this.entry,
    required this.builder,
  }) : super(key: key);

  /// [entry] is mandatory and should not be null
  /// assertion is used to check the nullability of the
  /// property and will show a stacktrace
  final List<Entry> entry;

  /// [builder] is mandatory and should not be null
  /// assertion is used to check the nullability of the
  /// property and will show a stacktrace
  final Widget Function(BuildContext context, int index, Entry item) builder;

  @override
  Widget build(BuildContext context) {
    return SliverList(
      delegate: SliverChildBuilderDelegate((context, index) {
        final item = entry[index];
        return builder(context, index, item);
      }, childCount: entry.length),
    );
  }
}

/// A wrapper above [_SliverHeaderDelegate]
///
class _SliverGroupedHeader extends StatelessWidget {
  const _SliverGroupedHeader({
    Key? key,
    required this.minHeight,
    required this.maxHeight,
    required this.child,
    this.pinned = false,
    this.floating = false,
  }) : super(key: key);

  final double minHeight;
  final double maxHeight;
  final Widget child;
  final bool pinned;
  final bool floating;

  @override
  Widget build(BuildContext context) {
    return SliverPersistentHeader(
        pinned: pinned,
        floating: floating,
        delegate: _SliverHeaderDelegate(
          minHeight: minHeight,
          maxHeight: maxHeight,
          child: child,
        ));
  }
}

/// A header delegate extending [SliverPersistentHeaderDelegate]
///
/// Will allow to display expendables Headers
class _SliverHeaderDelegate extends SliverPersistentHeaderDelegate {
  final double minHeight;
  final double maxHeight;
  final Widget child;

  _SliverHeaderDelegate({
    required this.minHeight,
    required this.maxHeight,
    required this.child,
  });

  @override
  double get minExtent => minHeight;
  @override
  double get maxExtent => max(maxHeight, minHeight);

  @override
  Widget build(
      BuildContext context, double shrinkOffset, bool overlapsContent) {
    return SizedBox.expand(child: child);
  }

  @override
  bool shouldRebuild(_SliverHeaderDelegate oldDelegate) {
    return maxHeight != oldDelegate.maxHeight ||
        minHeight != oldDelegate.minHeight ||
        child != oldDelegate.child;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sliver_grouped_list/sliver_grouped_list.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  static const double _kHeight = 150.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example'),
      ),
      body: SliverGroupedList<String, String>(
        data: {
          'A': ['My best friend', 'Good friend of mine', 'Guy I do not know'],
          'B': ['My cat', 'My dog', 'My fish', 'My bird'],
          'C': []
        },
        appBar: SliverAppBar(
          title: Text('SliverAppBar'),
          backgroundColor: Colors.green,
          expandedHeight: 200.0,
        ),
        header: SliverGrid.count(
          crossAxisCount: 3,
          children: [
            Container(color: Colors.white, height: _kHeight),
            Container(color: Colors.black, height: _kHeight),
            Container(color: Colors.grey, height: _kHeight),
            Container(color: Colors.grey, height: _kHeight),
            Container(color: Colors.white, height: _kHeight),
            Container(color: Colors.black, height: _kHeight),
            Container(color: Colors.black, height: _kHeight),
            Container(color: Colors.grey, height: _kHeight),
            Container(color: Colors.white, height: _kHeight),
          ],
        ),
        bodyHeaderMinHeight: 60.0,
        bodyHeaderMaxHeight: 100.0,
        bodyHeaderPinned: true,
        bodyHeaderFloating: false,
        bodyHeaderBuilder: (_, header) => Container(
          alignment: Alignment.center,
          child: Text(
            header,
            style: TextStyle(color: Colors.white),
          ),
          color: Colors.blue,
        ),
        bodyPlaceholderBuilder: (_, header) => Card(
            color: Colors.lightBlueAccent,
            child: Container(
                height: _kHeight,
                alignment: Alignment.center,
                child: Text(
                  "There are no items available in $header",
                  style: TextStyle(color: Colors.grey),
                ))),
        bodyEntryBuilder: (_, index, item) => GestureDetector(
          onTap: () {
            print(item);
            print(index);
          },
          child: Card(
              color: Colors.lightBlueAccent,
              child: Container(
                  height: _kHeight,
                  alignment: Alignment.center,
                  child: Text(
                    item,
                    style: TextStyle(color: Colors.white),
                  ))),
        ),
        footer: SliverFixedExtentList(
          itemExtent: _kHeight,
          delegate: SliverChildListDelegate(
            [
              Container(color: Colors.black),
              Container(color: Colors.white),
              Container(color: Colors.grey),
              Container(color: Colors.black),
              Container(color: Colors.white),
              Container(color: Colors.grey),
            ],
          ),
        ),
      ),
    );
  }
}
